`createTempFile` fails on linux with cgen error

### Example
```nim
import std/tempfiles
let (fda, patha) = createTempFile("", "")
```

### Current Output
seen in https://dev.azure.com/timotheecour/0c472305-f0f6-459d-aeab-b33734c4cb09/_apis/build/builds/3666/logs/72
```
2021-04-28T20:17:13.3476230Z CC: stdlib_tempfiles.nim
2021-04-28T20:17:13.4007854Z In file included from /usr/include/fcntl.h:289:0,
2021-04-28T20:17:13.4009355Z                  from /home/vsts/.cache/nim/testament_r/stdlib_tempfiles.nim.c:12:
2021-04-28T20:17:13.4010876Z In function ‘open’,
2021-04-28T20:17:13.4012305Z     inlined from ‘safeOpen_stdZtempfiles_7’ at /home/vsts/.cache/nim/testament_r/stdlib_tempfiles.nim.c:430:13:
2021-04-28T20:17:13.4016237Z /usr/include/x86_64-linux-gnu/bits/fcntl2.h:50:4: error: call to ‘__open_missing_mode’ declared with attribute error: open with O_CREAT or O_TMPFILE in second argument needs 3 arguments
2021-04-28T20:17:13.4017565Z     __open_missing_mode ();
2021-04-28T20:17:13.4018407Z     ^
2021-04-28T20:17:13.4137965Z In function ‘open’,
2021-04-28T20:17:13.4139158Z     inlined from ‘safeOpen_stdZtempfiles_7’ at /home/vsts/.cache/nim/testament_r/stdlib_tempfiles.nim.c:430:13,
2021-04-28T20:17:13.4140259Z     inlined from ‘createTempFile_stdZtempfiles_24’ at /home/vsts/.cache/nim/testament_r/stdlib_tempfiles.nim.c:585:25:
2021-04-28T20:17:13.4141733Z /usr/include/x86_64-linux-gnu/bits/fcntl2.h:50:4: error: call to ‘__open_missing_mode’ declared with attribute error: open with O_CREAT or O_TMPFILE in second argument needs 3 arguments
2021-04-28T20:17:13.4142543Z     __open_missing_mode ();
2021-04-28T20:17:13.4142862Z     ^
2021-04-28T20:17:13.4188632Z Error: execution of an external compiler program 'gcc -c  -w -fmax-errors=3 -O3 -fno-strict-aliasing -fno-ident   -I/home/vsts/work/1/s/lib -I/home/vsts/work/1/s/testament -o /home/vsts/.cache/nim/testament_r/stdlib_tempfiles.nim.c.o /home/vsts/.cache/nim/testament_r/stdlib_tempfiles.nim.c' failed with exit code: 1
2021-04-28T20:17:13.4192355Z 
```

### Expected Output
works

### Additional Information
1.5.1 486b1f621f2f4b3b02905db305d9a8d3985f06e4

--- a/lib/posix/posix.nim
+++ b/lib/posix/posix.nim
@@ -183,7 +183,11 @@ proc dlsym*(a1: pointer, a2: cstring): pointer {.importc, header: "<dlfcn.h>", s
 
 proc creat*(a1: cstring, a2: Mode): cint {.importc, header: "<fcntl.h>", sideEffect.}
 proc fcntl*(a1: cint | SocketHandle, a2: cint): cint {.varargs, importc, header: "<fcntl.h>", sideEffect.}
-proc open*(a1: cstring, a2: cint): cint {.varargs, importc, header: "<fcntl.h>", sideEffect.}
+proc openImpl(a1: cstring, a2: cint): cint {.varargs, importc: "open", header: "<fcntl.h>", sideEffect.}
+proc open*(a1: cstring, a2: cint, mode = 0.Mode): cint {.inline.} =
+  # prevents bug #17888
+  openImpl(a1, a2, mode)
+
 proc posix_fadvise*(a1: cint, a2, a3: Off, a4: cint): cint {.
   importc, header: "<fcntl.h>".}
 proc posix_fallocate*(a1: cint, a2, a3: Off): cint {.

--- a/lib/posix/posix.nim
+++ b/lib/posix/posix.nim
@@ -184,7 +184,7 @@ proc dlsym*(a1: pointer, a2: cstring): pointer {.importc, header: "<dlfcn.h>", s
 proc creat*(a1: cstring, a2: Mode): cint {.importc, header: "<fcntl.h>", sideEffect.}
 proc fcntl*(a1: cint | SocketHandle, a2: cint): cint {.varargs, importc, header: "<fcntl.h>", sideEffect.}
 proc openImpl(a1: cstring, a2: cint): cint {.varargs, importc: "open", header: "<fcntl.h>", sideEffect.}
-proc open*(a1: cstring, a2: cint, mode = 0.Mode): cint {.inline.} =
+proc open*(a1: cstring, a2: cint, mode: Mode | cint = 0.Mode): cint {.inline.} =
   # prevents bug #17888
   openImpl(a1, a2, mode)
 

--- a/lib/pure/streams.nim
+++ b/lib/pure/streams.nim
@@ -1503,6 +1503,7 @@ when false:
       of fmReadWrite: flags = O_RDWR or int(O_CREAT)
       of fmReadWriteExisting: flags = O_RDWR
       of fmAppend: flags = O_WRONLY or int(O_CREAT) or O_APPEND
+      static: doAssert false # handle bug #17888
       var handle = open(filename, flags)
       if handle < 0: raise newEOS("posix.open() call failed")
     result = newFileHandleStream(handle)

--- a/lib/std/tempfiles.nim
+++ b/lib/std/tempfiles.nim
@@ -8,6 +8,8 @@
 #
 
 ## This module creates temporary files and directories.
+##
+## Experimental API, subject to change.
 
 import os, random
 

--- a/lib/std/tempfiles.nim
+++ b/lib/std/tempfiles.nim
@@ -42,7 +42,7 @@ else:
   ): File {.importc: "fdopen",header: "<stdio.h>".}
 
 
-proc safeOpen(filename: string, mode: Mode): File =
+proc safeOpen(filename: string): File =
   ## Open files exclusively.
   # xxx this should be clarified; it doesn't in particular prevent other processes
   # from opening the file, at least currently.
@@ -66,6 +66,11 @@ proc safeOpen(filename: string, mode: Mode): File =
       discard close_osfandle(fileHandle)
       raiseOSError(osLastError(), filename)
   else:
+    # xxx we need a `proc toMode(a: FilePermission): Mode`, possibly by
+    # exposing fusion/filepermissions.fromFilePermissions to stdlib; then we need
+    # to expose a `perm` param so users can customize this (e.g. the temp file may
+    # need execute permissions), and figure out how to make the API cross platform.
+    let mode = Mode(S_IRUSR or S_IWUSR)
     let flags = posix.O_RDWR or posix.O_CREAT or posix.O_EXCL
     let fileHandle = posix.open(filename, flags, mode)
     if fileHandle == -1:
@@ -100,15 +105,10 @@ proc createTempFile*(prefix, suffix: string, dir = ""): tuple[fd: File, path: st
 
   createDir(dir)
 
-  # xxx we need a `proc toMode(a: FilePermission): Mode`, possibly by
-  # exposing fusion/filepermissions.fromFilePermissions to stdlib; then we need
-  # to add a `perm` param so users can customize this (e.g. the temp file may
-  # need execute permissions).
-  let mode = Mode(S_IRUSR or S_IWUSR)
   for i in 0 ..< maxRetry:
     result.path = dir / (prefix & randomPathName(nimTempPathLength) & suffix)
     try:
-      result.fd = safeOpen(result.path, mode)
+      result.fd = safeOpen(result.path)
     except OSError:
       continue
     return

--- a/lib/std/tempfiles.nim
+++ b/lib/std/tempfiles.nim
@@ -42,8 +42,10 @@ else:
   ): File {.importc: "fdopen",header: "<stdio.h>".}
 
 
-proc safeOpen(filename: string): File =
+proc safeOpen(filename: string, mode: Mode): File =
   ## Open files exclusively.
+  # xxx this should be clarified; it doesn't in particular prevent other processes
+  # from opening the file, at least currently.
   when defined(windows):
     let dwShareMode = FILE_SHARE_DELETE or FILE_SHARE_READ or FILE_SHARE_WRITE
     let dwCreation = CREATE_NEW
@@ -65,8 +67,7 @@ proc safeOpen(filename: string): File =
       raiseOSError(osLastError(), filename)
   else:
     let flags = posix.O_RDWR or posix.O_CREAT or posix.O_EXCL
-
-    let fileHandle = posix.open(filename, flags)
+    let fileHandle = posix.open(filename, flags, mode)
     if fileHandle == -1:
       raiseOSError(osLastError(), filename)
 
@@ -93,17 +94,21 @@ proc createTempFile*(prefix, suffix: string, dir = ""): tuple[fd: File, path: st
   ## If failing to create a temporary file, `IOError` will be raised.
   ##
   ## .. note:: It is the caller's responsibility to remove the file when no longer needed.
-  ##
   var dir = dir
   if dir.len == 0:
     dir = getTempDir()
 
   createDir(dir)
 
+  # xxx we need a `proc toMode(a: FilePermission): Mode`, possibly by
+  # exposing fusion/filepermissions.fromFilePermissions to stdlib; then we need
+  # to add a `perm` param so users can customize this (e.g. the temp file may
+  # need execute permissions).
+  let mode = Mode(S_IRUSR or S_IWUSR)
   for i in 0 ..< maxRetry:
     result.path = dir / (prefix & randomPathName(nimTempPathLength) & suffix)
     try:
-      result.fd = safeOpen(result.path)
+      result.fd = safeOpen(result.path, mode)
     except OSError:
       continue
     return

